File: Core/src/com/forcex/anim/Animation.java
package com.forcex.anim;
import java.util.*;

public class Animation
{
	float duration = 0;
	ArrayList<Bone> bones = new ArrayList<Bone>();
	public String name = "";

	public Animation(String name){
		this.name = name;
	}

	public void addBone(Bone bone){
		bones.add(bone);
	}

	public Bone findBone(int index){
		for(int i = 0;i < bones.size();i++){
			if(bones.get(i).boneID == index){
				return bones.get(i);
			}
		}
		return null;
	}

	public float getDuration(){
		if(duration == 0){
			for(Bone bone : bones){
				for(KeyFrame key : bone.keyframes){
					if (key.time > duration) {
						duration = key.time;
					}
				}
			}
		}
		return duration;
	}
	
	public Bone getRootBone(){
		return bones.get(0);
	}
}

File: Backend/android-backend/app/src/main/java/com/forcex/anim/Animation.kt
package com.forcex.anim

import java.util.*

class Animation(name: String) {
    internal var duration = 0f
    internal var bones = ArrayList<Bone>()
    var name = ""

    val rootBone: Bone
        get() = bones[0]

    init {
        this.name = name
    }

    fun addBone(bone: Bone) {
        bones.add(bone)
    }

    fun findBone(index: Int): Bone? {
        for (i in bones.indices) {
            if (bones[i].boneID == index) {
                return bones[i]
            }
        }
        return null
    }

    fun getDuration(): Float {
        if (duration == 0f) {
            for (bone in bones) {
                for (key in bone.keyframes) {
                    if (key.time > duration) {
                        duration = key.time
                    }
                }
            }
        }
        return duration
    }
}
